tests/testthat/test_Plot_synergy.R

```diff
@@ -0,0 +1,214 @@
+test_that("color_ranges_from_green_to_white_to_red", {
+  expected_colors <-
+    c("#00FF00", "#3FFF3F", "#7FFF7F", "#BFFFBF", # 00ff00 is green
+      "#FFFFFF", "#FFBFBF", "#FF7F7F", "#FF3F3F", # ffffff is white
+      "#FF0000")                                  # ff0000 is red
+  expect_equal(expected_colors, .GetColors(-8,8))
+})
+
+test_that("slice_at_least_first_works", {
+  expect_equal(2:5, .SliceAtLeastFirst(c(-1,5)))  # pathological case
+  expect_equal(2:5, .SliceAtLeastFirst(c(0,5)))   # degenerate case
+  expect_equal(2:5, .SliceAtLeastFirst(c(1,5)))
+  expect_equal(2:5, .SliceAtLeastFirst(c(2,5)))
+  expect_equal(3:5, .SliceAtLeastFirst(c(3,5)))
+  expect_equal(3:5, .SliceAtLeastFirst(c(3,5,10)))  # spurious val 10
+})
+
+test_that("numeric_apply_range_works", {
+  val1 <- c("1", 2, 3)
+  val2 <- c('2', 4, '6', 8, '10')
+  expect_equal(c(1,2,3), .NumericApplyRange(val1, NULL))
+  expect_equal(c(1,2,3), .NumericApplyRange(val1, c(1,3)))
+  expect_equal(c(1,2,3), .NumericApplyRange(val1, c(1,3,5))) # spurious val 5
+  expect_equal(c(2,4,6,8,10), .NumericApplyRange(val2, NULL))
+  expect_equal(c(4,6,8), .NumericApplyRange(val2, c(2,4)))
+})
+
+test_that("apply_ranges_works", {
+  data <- matrix(c(1,2,3,
+                   4,5,6,
+                   7,8,9), nrow = 3, ncol = 3, byrow=TRUE)
+
+  # Various ways to apply ranges that change nothing.
+  expect_equal(data, .ApplyRanges(data, NULL, NULL))    # no ranges
+  expect_equal(data, .ApplyRanges(data, c(1,3), NULL))  # full row range
+  expect_equal(data, .ApplyRanges(data, NULL, c(1,3)))  # full col range
+  expect_equal(data, .ApplyRanges(data, c(1,3), c(1,3))) # both full ranges
+
+  # Trim just rows or just columns.
+  expect_equal(matrix(c(1,2,3,
+                        4,5,6), nrow=2, ncol=3, byrow=TRUE),
+               .ApplyRanges(data, c(1,2), NULL))
+  expect_equal(matrix(c(1,2,
+                        4,5,
+                        7,8), nrow=3, ncol=2, byrow=TRUE),
+               .ApplyRanges(data, NULL, c(1,2)))
+
+  # Trim both rows and columns.
+  expect_equal(matrix(c(1,2,
+                        4,5), nrow=2, ncol=2, byrow=TRUE),
+               .ApplyRanges(data, c(1,2), c(1,2)))
+})
+
+test_that("extended_scores_works", {
+  # Note that we expect the kriging in this simple case to preserve
+  # the corner values 1,2,3,4.  In between, we expect to see additional
+  # interpolated data points that show a progression from one original
+  # value to the next.
+  data <- matrix(c(1,2,
+                   3,4), nrow=2, ncol=2, byrow=TRUE)
+  print(c(data))
+  extended_data <- matrix(c( 1.00, 1.33, 1.69, 2.00,
+                             1.65, 1.98, 2.33, 2.65,
+                             2.36, 2.69, 3.05, 3.36,
+                             3.00, 3.32, 3.68, 4.00),
+                          nrow=4, ncol=4, byrow=TRUE)
+  expect_equal(extended_data, .ExtendedScores(data, 2), tolerance=1e-02)
```

Comment:
Please note this unit test.  I start with a square matrix
[1.2.
.3.4]

I expect that, after kriging, those four numbers ought to stay in the corners they currently occupy.

Is that not correct?

Reply:
Yeah, it's correct. With the new .ExtendedScores function, we need a non-square matrix example here, thanks.

Reply:
All I did for this was to run the function, and write down its output as the initial unit test.

"Isn't that bogus?  What if the function you used to generate that output was buggy?",  you might ask.  You would be right!  It is best to have independently known-good data to serve for comparison.  However the bogus test is not totally without value:  If you change the implementation in such a way that the output changes, even a test with bogus comparison data will fail, signaling the output change.  Then you can explicitly decide whether the change was deliberate (and update the test) or undesired (and fix the code).

Anyway, per the discussion above with your .zip file for .ExtendedScores(), I suggest that you follow up my submit with a small one of your own, that fixes up the implementation and also adds a unit test case.

Does that sound OK to you?